ARROW-12150: [Python] Correctly infer type of mixed-precision Decimals

jorisvandenbossche opened the pull request:
We were currently taking the max of current + new precision and scale, independently. But a new value could require a higher precision if it's number of integrals (precision - scale) is higher than the current set.

jorisvandenbossche commented on cpp/src/arrow/python/decimal.cc:
```diff
@@ -214,12 +214,14 @@ DecimalMetadata::DecimalMetadata(int32_t precision, int32_t scale)
   scale_ = std::max(current_scale, suggested_scale);
 
+  const int32_t current_precision = precision_;
+  auto n_integral = std::max(std::max(current_precision, 0) - std::max(current_scale, 0),
```
Those additional `max(current, 0)` checks make it a bit ugly, but this is required (I think) because on the first call, the current_precision or current_scale can still be the int32 min value (-2147483648)

cyb70289 replied:
If `current_precision` is int32:min, looks `current_scale` must also be int32::min? So without std::max, guess it should also work.
But I think keep std::max there is better in case we change the behaviour.

pitrou commented on python/pyarrow/tests/test_convert_builtin.py:
```diff
@@ -1501,6 +1501,23 @@ def test_sequence_decimal_too_high_precision():
+         pa.decimal128(6, 3)),
+        ([decimal.Decimal('123.4'), decimal.Decimal('4.567')],
+         pa.decimal128(6, 3)),
+    ]
```
What about e.g. `[Decimal('123e2'), Decimal('4567e3')]`?

jorisvandenbossche replied:
Ah, yes, I didn't think about negative scale ..

pitrou commented:
So here is how inference should work IMHO:
* compute both the magnitude of most-significant digit and least-significant digit
* examples:
  * `Decimal(12345) => msd = 4, lsd = 0`
  * `Decimal(12345e2) => msd = 6, lsd = 2`
  * `Decimal(1234.5) => msd = 3, lsd = -1`,
  * `Decimal(0.012345) => msd = -2, lsd = -6`
  * `Decimal(0) => msd = -inf, lsd = +inf`
* track `max(msd)` and `min(lsd)`
* at the end, `precision = max_msd - min_lsd + 1`, `scale = -min_lsd`

jorisvandenbossche commented:
@pitrou thanks a lot for that input.

To be explicit: do you think this logic should be used for *all* inference code paths? 

Because currently, our inference of scale/precision from a string (`Decimal128::FromString`) vs from a python decimal (`InferDecimalPrecisionAndScale`)  have some different behaviour regarding trailing zero's. For a value like `Decimal('123E+2')`, the first would give a (precision, scale) of (5, 0), while the other (3, -2). 
And it is this difference that gives a problem because now we first infer the type from the python decimal in a first pass, but then for the actual conversion in a second pass go through the string representation of the decimals, which gives conflicting types, even in the case of non-mixed decimals. For example with a single decimal (on master):

```
In [1]: from decimal import Decimal

In [2]: pa.array([Decimal('123E+2')])
...
ArrowInvalid: Decimal type with precision 5 does not fit into precision inferred from first array element: 3
../src/arrow/python/python_to_arrow.cc:169  internal::DecimalFromPyObject(obj, *type, &value)
../src/arrow/python/python_to_arrow.cc:486  PyValue::Convert(this->primitive_type_, this->options_, value)
../src/arrow/python/iterators.h:69  func(value, static_cast<int64_t>(i), &keep_going)
../src/arrow/python/python_to_arrow.cc:1055  converter->Extend(seq, size)

In [3]: pa.infer_type([Decimal('123E+2')])
Out[3]: Decimal128Type(decimal128(3, -2))
```

I think the logic you proposes matches the current logic of inferring from a python decimal, but thus not when inferring from a string.

pitrou commented:
Well, the second pass shouldn't do any inference at all. Instead there probably should be a `Decimal128::FromString(util::string_view, int scale)` for fixed-scale conversion.

pitrou commented:
@jorisvandenbossche Do you plan to continue working on this?

pitrou commented on python/pyarrow/tests/test_convert_builtin.py:
```diff
@@ -1501,6 +1501,42 @@ def test_sequence_decimal_too_high_precision():
+        ([decimal.Decimal('123.4'), decimal.Decimal('4.567')],
+         pa.decimal128(6, 3)),
+        ([decimal.Decimal('123e2'), decimal.Decimal('4567e3')],
+         pa.decimal128(5, -2)),
```
Hmm... can you add a test for `[decimal.Decimal('123e4'), decimal.Decimal('4567e2')]`, and also one for `[decimal.Decimal('0.123'), decimal.Decimal('0.04567')]`?

pitrou commented on python/pyarrow/tests/test_convert_builtin.py:
```diff
@@ -1501,6 +1501,42 @@ def test_sequence_decimal_too_high_precision():
+        (decimal.Decimal('12300'), pa.decimal128(5, 0)),
+        (decimal.Decimal('12300.0'), pa.decimal128(6, 1)),
+        # scientific power notation
+        (decimal.Decimal('123E+2'), pa.decimal128(3, -2)),
```
What about `decimal.Decimal('0.0123')`?

pitrou reviewed (commented):
I'd like to see this a bit more tested so that we're sure to get it right, see below.

jorisvandenbossche commented:
For now, I pushed a minimal fix to ensure we don't raise for a different precision if we rescale the decimal. That at least fixes the existing error for eg `pa.array([Decimal('123E+2')])`.

But still running into some some more corner cases regarding the precision/scale inference from a Python Decimal.

Using released version of pyarrow, we have an inconsistency in how to deal with positive exponents:

```python
>>> pa.infer_type([decimal.Decimal('123e4')])
Decimal128Type(decimal128(7, 0))
>>> pa.infer_type([decimal.Decimal('123e2')])
Decimal128Type(decimal128(3, -2))
```

Those two cases should be interpreted consistently, and following https://github.com/apache/arrow/pull/9948#issuecomment-817835991 they should give `(3, -4)` and `(3, -2)`. The reason for this difference is this `num_digits <= abs_exponent` check (so "123e4" with the exponent of 4 being larger than number of digits of 3 is handled differently):

https://github.com/apache/arrow/blob/b3e43987c47b2f01b204a2d954f882f7161616ef/cpp/src/arrow/python/decimal.cc#L79-L84

This check is to ensure to add precision in case of leading zeros as in `decimal.Decimal('0.0123')`:

```python
>>> decimal.Decimal('0.0123').as_tuple()
DecimalTuple(sign=0, digits=(1, 2, 3), exponent=-4)
>>> pa.infer_type([decimal.Decimal('0.0123')])
Decimal128Type(decimal128(4, 4))
```
So we could do this `num_additional_zeros` edit only when the exponent is negative, to ensure that `Decimal('123e4')` and `Decimal('123e2')` (both with a positive exponent) are treated consistently. 

However, changing that would change the tested behaviour for:

```python
>>> pa.infer_type([decimal.Decimal('0.01E5')])
Decimal128Type(decimal128(4, 0))
```

to change to `decimal128(1, -3)`. Are we fine with changing this tested behaviour for `decimal.Decimal('0.01E5')]` ?

pitrou commented:
I would be fine with making the suggested change. I would also prefer if the inference logic remained simple and easy to understand :-)

jorisvandenbossche commented:
The inference logic itself actually became simpler with the current state of the PR, IMO ;) (only the update-logic for combining multiple inferred types got a bit more complicated, but that's the necessary bug fix)

Now, whether to interpret `Decimal('123e2')` as `decimal128(5, 0)` or `decimal128(3, -2)`: this PR changes that to `(3, -2)`, but there might also be some arguments to not use negative scales by default in case of ambiguity. Many SQL engines don't actually support negative scales, and also Parquet requires the scale to be 0 or a positive integer < precision (https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#decimal).

pitrou commented:
Thanks for the precision ( ;-) ) @jorisvandenbossche . If negative scale is not well supported everywhere, then I agree it makes sense to avoid it.

pitrou commented:
@jorisvandenbossche Do you want to update this PR?

jorisvandenbossche commented:
OK, updated this, and the last commit changes the default inference from a python Decimal (`InferDecimalPrecisionAndScale`) to not use a negative scale (you can of course still specify a type manually with a negative scale to force it). 

The last commit only contains that change (with changes in the tests to follow the new behaviour). Additionally, I also added a python test to check several cases using a specified type (just to cover that case as well, it should not be impacted by the changes in the last commit)

pitrou reviewed (approved):
Thanks for the update @jorisvandenbossche !

jorisvandenbossche commented:
And thanks for the simplification! That's better ;)
